django/contrib/admin/sites.py

```diff
@@ -245,20 +247,23 @@ def wrapper(*args, **kwargs):
 
         # Add in each model's views, and create a list of valid URLS for the
         # app_index
-        valid_app_labels = []
+        valid = defaultdict(set)
         for model, model_admin in self._registry.items():
             urlpatterns += [
                 url(r'^%s/%s/' % (model._meta.app_label, model._meta.model_name), include(model_admin.urls)),
             ]
-            if model._meta.app_label not in valid_app_labels:
-                valid_app_labels.append(model._meta.app_label)
+            valid['app_label'].add(model._meta.app_label)
+            valid['model_name'].add(model.__name__)
 
         # If there were ModelAdmins registered, we should have a list of app
         # labels for which we need to allow access to the app_index view,
-        if valid_app_labels:
-            regex = r'^(?P<app_label>' + '|'.join(valid_app_labels) + ')/$'
+        if valid['app_label']:
+            valid_app_label_regex = r'(?P<app_label>' + '|'.join(valid['app_label']) + ')'
+            valid_model_name_regex = r'(?P<model_name>' + '|'.join(valid['model_name']) + ')'
             urlpatterns += [
-                url(regex, wrap(self.app_index), name='app_list'),
+                url(r'^%s/$' % valid_app_label_regex, wrap(self.app_index), name='app_list'),
+                url(r'^autocomplete/%s/%s/(?P<field_name>.+)/$' % (valid_app_label_regex, valid_model_name_regex),
+                    wrap(AutocompleteJsonView.as_view(admin_site=self)), name='autocomplete'),
```

Comment:
Couldn't this be constrained to just the models that actually define autocomplete fields in their admins? Ostensibly `get_autocomplete_fields` takes a `request` parameter but it [doesn't actually use it](https://github.com/codingjoe/django/blob/2763841ea2935b6c02ba462eaff5106da667d1e8/django/contrib/admin/options.py#L284)